### DISCUSSION TITLE
Positron IDE could have a better layout

### DISCUSSION THREAD

**Discussion by justinasR-19** [posit-dev/positron#5853] [Feedback]
I love the Positron IDE already with its all new features. It does seem like an IDE for data scientists. 

When I got myself into the data science field, I spent countless hours trying to find my workflows. One of those where searching for the IDE that I feel the most productive. In the end I stayed with Rstudio. Its not perfect, in many areas, but.... what truly works for me in Rstudio that didnt work in other popular IDE is the layout. 

What I mean layout is..... everything is compact. None of the buttons are oversized, unlike in VScode, the spacing between items in the environamnet is gret in Rstudio, no side panels on the left that only takes space, everything strict and in order. Console header doesnt take 25% of the screen height, unlike in VSCode and so on....

![image](https://github.com/user-attachments/assets/1d93f30d-e203-4294-85ff-a6f0a3192cf2)

Positron does look to inherit the style from VScode, which I absolutely hate. My hope is that you guys make Positron look more like Rstudio. All I want is just my screen space.

**Comment by jmcphers** [comment]
Positron is designed to feel familiar to users of both VS Code and RStudio, so its UI lands somewhere between those two systems. If you want it to look more like the RStudio screenshot you shared, you can rearrange it a bit:

- Right-click on the sidebar go to _Activity Bar Position_ and select _Hidden_
- Drag the Console over to the right quadrant

Here's what that looks like:

<img width="1349" alt="image" src="https://github.com/user-attachments/assets/8ffc14d1-a789-4946-b517-217c1bb8085a" />

**Reply by justinasR-19** [reply]
i tinkered around and i managed to find a layout I like in Positron. The VSCode is not bad, i just struggle with the space wasted. Such as too much spacing for the ribbon at the bottom where TERMINAL button sits. Its possible to squeeze another millimeter or two of screen area by making it not as tall. Obviously, im being bias here :) I like the design of RStudio and the functionality of Positron. Either way, great job on the product.
Many thanks.

**Comment by ntluong95** [comment]
I think you are quite conservative. What I truly value Posit team is while they utilize CodeOSS (a VSCode like) with a lot of extensibility and flexibility, they try to preserve Rstudio experience. What is the point of having a new product which just look similar to the old one? Then they should have just focussed on developing Rstudio

**Reply by Jiayou-Chao** [reply]
Because you will be able to leverage the VSCode ecosystem.

**Reply by justinasR-19** [reply]
> I think you are quite conservative. What I truly value Posit team is while they utilize CodeOSS (a VSCode like) with a lot of extensibility and flexibility, they try to preserve Rstudio experience. What is the point of having a new product which just look similar to the old one? Then they should have just focussed on developing Rstudio

Yes, you are right. Im very conservative in the IDE world. I guess ill have to get used to the new layout

**Comment by TommasoGhilardi** [comment]
I share many of the same sentiments as @justinasR-19 . While I understand that making Positron too similar to RStudio might be counterproductive (since you could simply use RStudio in that case), RStudio really nails a few core design features that would be great to see implemented in Positron.

Of course, these are just my personal feelings, but for me there’s a lot of wasted space—e.g. in the header of the VARIABLES, PLOT etc sections—which reduces the area available for the information I really care about. I tried lowering the font size, but that doesn’t fully address the issue.

Another challenge is the limited flexibility in arranging panels. Currently, sections (variables, plots, console, etc.) can only be folded, and once folded, everything else shifts up or down. It would be incredibly helpful to have an option for multiple section in the same panel—similar to RStudio—where I can switch views in a single click without anything moving around.

All of this is to say: I love Positron, and I’d love to see it adopt some of the aspects that made me appreciate RStudio so much.

**Comment by juliasilge** [comment]
I opened https://github.com/posit-dev/positron/issues/5885 to explore whether we can reduce the number of toolbars on the Variables pane from two to one.

The UI infrastructure we are building on here is actually much more flexible than RStudio's, where there are only four "spots" that you can plonk content into. I'll highlight some of the options you have here:

![Screenshot 2025-01-06 at 11 55 28 AM](https://github.com/user-attachments/assets/92b8457f-f70c-4c5b-b3bd-d3f76efc47c3)

If you don't want to see, for example, the status bar at the bottom, you can turn it off from this menu.

**Reply by TommasoGhilardi** [reply]
Thank you so much, @juliasilge ! Streamlining the interface would definitely make it more appealing—though I understand it might not be a top priority at the moment. I do realize the UI is already more flexible than RStudio in some ways, but what I was referring to is the ability in RStudio to stack panels horizontally (like in the screenshots). Personally, I prefer that kind of layout.

Right now, if I want to focus on a single panel and minimize another one (for example, minimizing the plot to expand the variables), all the panels move up resizing. Having an option for this “horizontal stacking” would really help imo.
![image](https://github.com/user-attachments/assets/51a7de90-027b-4330-83f9-7eaaa81c247b)

**Comment by Justinas365** [comment]
What I mean with with initial post about making it look like RStudio (making everything compact) instead of VSCode is as per below:
![one](https://github.com/user-attachments/assets/2f2d44ba-4a4d-4ea7-bd27-6fa7053fa80c)

1 - there is no need for this bar to be so wide. Those icons could be smaller. I would appreciate more real estate on the scripting window.
2 - I feel that there is too much spacing above and below. I would appreciate more real estate on my scripting window
3 - This ribbon is again, too tall for my eye. The fact that I have "CONSOLE" bar at the top, I would perhaps like to combine those two. I can see at least in another 2 places, which version of R im using. Perhaps I dont wnat it here, or perhaps I want it here but i dont want it somewhere else.
4 - Again, its just too tall for my eye. Wasted space. the fact both sections are coming under "VARIABLE" section, I would like it so see organised somehow different. For example. If i collapse "VARIABLE", I would want to know the bottom sections also going to collapse.
5 - I see in another two places which R version im using.
6 - Filtering could be moved somewhere else, because it takes so much space. really wasted space here
7 - The bar is too tall for my eye. 
8 - Tabs are way too tall for my eye, I would appreciate more real estate for my scripting window
9 - 50% of this bar is emtpy space and its quite tall. i have read what this search window is, but I would rather go with CTRL+F or CTRT+SHIRT+F (I think it searches the whole project in RStudio). The rest of the buttons could be incorporated somewhere else and getting back that scripting real estate would be much appreciated.
9 - Looks like I duplicated nines :) i would want this to be much narrower and get my scripting real estate back. you know the drill be now haha. 
10 - What i really miss from RSstudio is the STOP button. That is just iconic. And you know if the code is working or now. R is not R without that button in my head. Python can benefit from that button as well.
EDIT: I just noticed that green line in the console, which is running the code. Which looks nice. I think it even looks better than that stop sign. But it would be nice to have a way to kill the process, like that stop sign allowed.
11 - I like it, but I would happy to remove it if this helps to improve space somewhere else.

BIG ISSUE: The whole yellow area. there are too many sections, too many lines too many shades (this could also be an issue of cobalt theme). What I loved about RStudio is the tab layout below:
![three](https://github.com/user-attachments/assets/b61c06e2-3f15-476d-888b-1b0a3c2765eb)


![two](https://github.com/user-attachments/assets/54d8b5a0-6021-457e-a13a-14432ed89304)
13 - This bar is really not needed. I would appreciate to have more viewing space to understand my data
14 - Too wide spacing. I would appreciate to have some more viewing data real estate. What I havent highlighted is the histograms of each variable. Its actually very nice, but perhaps too big for my liking. 
EDIT: In fact, I think the histograms can be moved to columns under each name. Simply, because the side bar is replicating most of the data already (col name, type). In this case, I could see so many more columns.
15, 19, 20 - For me the font is off. I dont know what is it. My eyes get tired looking at it. Perhaps its too thick, perhaps its too bold. the whole window is creaming in my eyes and I dont know where to look.  My eyes are really liking the design highlighted in yellow.
16 - I havent checked the documentation what this does, but if this somehow related to the number on its left, can we get rid of it? i would appreciate more real estate for data viewing.
17 - This button currently only shows up when my cursor is around. But it should be there, patiently waiting to meet my cursor continuously.
18 - I think a simple line would do. I would appreciate more real estate to view my data.

RStudio had an excellent viewer. I am able to see more than 3 columns of my data, the data values are not bolded, so its not screaming in my face. In fact, looking at this window makes me calm. So this is a big help when the work is stressful :) 
 
![four](https://github.com/user-attachments/assets/743c0a6d-e7eb-4029-856e-dc57a7348d88)


Bottom line: obviously I bashed Positron left and right here, without saying anything good about it. But in fact. I do like it very much. Really appreciate your work and I truly want this to become my go to. RStudio was not ideal, but it was good enough for me to choose it as my main IDE. Some things can and should be brought to Positron. 

P.S. Could you please make COBALT theme as native Positron? You nailed it in RStudio. the one i found on the market place is a bit too much for my eyes.

Many thank, Justinas

**Reply by ramiromagno** [reply]
@Justinas365: Thank you so much for the write-up! I totally agree with all your points. Thank you for pointing out all the reasons why I haven't switched to Positron yet.

**Comment by ntluong95** [comment]
There are many points here and of course I am not the right one to answer, but here is some of my personal thoughts:
- Spacing: that looks reasonable to my eye and cannot smaller. Reducing 1 unit of the margin will not bring any space for you to display code. Better to have a 2nd wide screen 😄
- Arrangement of some buttons: I agree some of them could be rearrange to bring more space. But it should be also balance between space and simplicity. Cluttering many things in the same line make it hard to navigate
- Actually the VSCode layout brings you with more flexibility. You can drag windows/views to everywhere, collapse or extend them. In Rstudio, you do it by going to to Global Options/ Pane Layout. It works if you don't usually change your layout or workflow, but Positron/VSCode is for multi-purpose workflow. I use both of them so I dont want Positron change too much Positron from VScode as I am maintaining a single setting.json file to make them look identical. But having a stack viewed like Rstudio for both top and bottom right pane would be perfect. That is the only thing I need but I know it might be hard for them to implement.

**Reply by Justinas365** [reply]
i am often working on 14inch laptop, so millimeter here and there helps. I guess the main point about those "millimeters" is that the bars or ribbons (what ever those are called) is just too tall for my eyes. What my eyes are truly in love is the first screenshot the very bottom line, where its says main branch, quarto 1.5 bla bla bla. the bar is not tall, my eyes can read clearly whats written and i dont get lost among too many spacing and lines. I used to used VSCode, but I could never get used to its design, hence moved to RStudio

**Comment by ZhimingYe** [comment]
The device I commonly use is a 13.3-inch MacBook Air. Before Positron was developed, I frequently used the Jupyter functionality in VSCode for Python and RStudio for R, so I am very familiar with both tools. Speaking from my personal perspective, elements like the height of the tab bar, the height of lists, and the size of the data browser grid cells are indeed issues that need to be considered on small-screen devices. I support making these customizable options.

Compared to RStudio, for presenting content on small screens, the default heights are truly a disaster and far from ideal. In this regard, I completely agree with @Justinas365 and @TommasoGhilardi ’s opinion.

However, for Positron, I believe the interface should not be a primary focus, as the team has a significant number of functionality issues awaiting resolution. This could perhaps be treated as a low-priority issue to be addressed later.

This is a simple example. Rstudio provides more info on screen: 
![Example](https://github.com/user-attachments/assets/9e196454-f83b-40eb-8586-14f7011485db)

**Reply by juliasilge** [reply]
Specifically on that summary panel for the Data Explorer, you may be interested in adding a 👍 to https://github.com/posit-dev/positron/issues/5435 and/or https://github.com/posit-dev/positron/issues/4370.

**Reply by ZhimingYe** [reply]
> Specifically on that summary panel for the Data Explorer, you may be interested in adding a 👍 to #5435 and/or #4370.

Oh, my point is not the summary panel but the **height** of each cell (both the data explorer or the environment variable panel). But this is not an important issue 😃. Apologize for misleading.

**Comment by fabeit** [comment]
Hello,
For me one of the most frustrating things is the management of multiple windows. It seems Positron is not designed to work on multiple monitors. I have to MANUALLY pop up the code editor every time I open a new file. Positron insists on opening a file always in the main window, unless I use the open file menu. So at each startup is the same chore. Also the management of the plots and the help file all opening in the main window take up so much space.

I am wondering, how big of a monitor do you have?! On a single monitor all these windows become so small I can barely read any text. Sorry but I don't understand this one-monitor approach when most programmers have at least 2 monitors.

It's a shame because I like many things about Positron but the UI is clunky.
